### DISCUSSION TITLE
Create a Gist for you script and it's lib dependencies

### DISCUSSION THREAD

**Discussion by BeSpunky** [johnlindquist/kit#1246] [Share]
## Watch how I publish the script that publishes scripts and their dependencies to Gist... 😄

https://user-images.githubusercontent.com/47897671/236570763-9c84163f-c4f3-46d6-943f-537724db2b2e.mp4

Here's the Gist of it:
https://gist.github.com/BeSpunky/ff5dcb62887cbee686dd6c3ba31cabb5

**TLDR**
As I go playing with ScriptKit, I started using the `lib` folder to centralize reusable functionality.
This made my scripts difficult to share, as they have nested dependencies which I would've had to add manually to my Gists.
Well no more... 💪

This script let's you choose one of your scripts, reads it and recursively extracts `lib` dependencies, then publishes a new gist with the script and the dependencies.

Enjoy 😊

**Comment by Schmedu** [comment]
That's a nice way to share a script, but the import is still a PITA 😅
An alternative approach would be to just share the generated `.mjs` file in the `~/.kenv/.scripts` folder as they include already all necessary imported functions.

**Reply by BeSpunky** [reply]
By import you mean installing the scripts? Yes. It isn't straightforward. At the moment this is more a "learn from it and take the parts you need" kind of share (unless you're willing to manually figure out where things go and put them there).

I discussed making these installable with John on Discord. He suggested trying the gist API. I'll be working on that sometime in the near future.

Kenvs were another way of sharing code that was mentioned. They are installable, bet they require a dedicated repo, which might be inconvenient if you just want to share a small piece of code with the community.

Sharing the `.mjs` means losing types formatting and comments. It beats the purpose of sharing IMO.

**Reply by Schmedu** [reply]
> I discussed making these installable with John on Discord. He suggested trying the gist API. I'll be working on that sometime in the near future.

Nice 👍

> Sharing the .mjs means losing types formatting and comments. It beats the purpose of sharing IMO.

IMO the purpose of sharing is to have someone else actually install and use your script - in the easiest way possible. But I agree with you that preserving types, formatting and comments would be preferable.

Nonetheless, thanks for the nice script! I was playing around with a similar thought, good that you implemented it already 🙌🏼

**Reply by johnlindquist** [reply]
I've been thinking a lot about the concept of "installer scripts". (Needs a better name 😅)

The general idea being that we would need to author an "installer script" that would point to the proper gist/resources/etc. And then you share the installer script which clearly presents to the user what it's going to do.

Installer scripts could live in their own "bundles"/whatever dir so they don't show up in the main "scripts" dir.

Anyway, that's my current thinking. It would definitely require some conventions/examples for it to make sense.

**Reply by BeSpunky** [reply]
Sure thing. My pleasure. 😊 
And thanks for the nice words.

These indeed must be installable. I'll make sure to share the installation script once it's ready.

**Reply by BeSpunky** [reply]
> I've been thinking a lot about the concept of "installer scripts". (Needs a better name 😅)
> 
> The general idea being that we would need to author an "installer script" that would point to the proper gist/resources/etc. And then you share the installer script which clearly presents to the user what it's going to do.
> 
> Installer scripts could live in their own "bundles"/whatever dir so they don't show up in the main "scripts" dir.
> 
> Anyway, that's my current thinking. It would definitely require some conventions/examples for it to make sense.

Definitely a good idea.
I'm imagining a new package with "installation" tools. Something between these lines:
```ts
import { installGist, InstallerConfig } from '@johnlindquist/kit/installer';

const gistUrl = 'https://whatever......';
const installerConfig: InstallerConfig = {
    existingFiles: 'overwrite', // or 'backupAndReplace' or 'merge' ? 😅
    showWarning: true,
    warningHtml: (files) => `..........`
};
const installation = await installGist(gistUrl, installerConfig); // Will show the default or custom warning

if (installation.success)
{
    inspect(installation.files); // { path: string, result: 'created' | 'overwritten' | 'merged' }[]
}
```

Are you thinking of something similar?

**Reply by johnlindquist** [reply]
Yeah, precisely 👍

**Comment by Schmedu** [comment]


**Reply by johnlindquist** [reply]
Kenvs are awesome for version control/etc.

Installers could handle things like "install all deps/ffmpeg", kinda like npm's postinstall.

It will definitely depend on the goals/intentions of the script author
